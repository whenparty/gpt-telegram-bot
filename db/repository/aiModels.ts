export const enum AI_MODEL {
  CLAUDE_3_HAIKU = "claude-3-haiku",
  CLAUDE_3_OPUS = "claude-3-opus",
}

export const AI_MODEL_API_VERSION = {
  [AI_MODEL.CLAUDE_3_HAIKU]: "claude-3-haiku-20240307",
  [AI_MODEL.CLAUDE_3_OPUS]: "claude-3-opus-20240229",
};
