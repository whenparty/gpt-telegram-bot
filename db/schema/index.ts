export * from "./users";
export * from "./tokens";
export * from "./messages";
export * from "./enums";
