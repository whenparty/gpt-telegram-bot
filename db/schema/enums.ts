import { AI_MODEL } from "db/repository/aiModels";
import { pgEnum } from "drizzle-orm/pg-core";

export const aiModelEnum = pgEnum("ai_model", [
  AI_MODEL.CLAUDE_3_HAIKU,
  AI_MODEL.CLAUDE_3_OPUS,
]);
