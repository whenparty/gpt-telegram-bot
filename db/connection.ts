import { Pool } from "pg";
import { drizzle } from "drizzle-orm/node-postgres";
import * as schema from "./schema";

export const connection = new Pool({
  connectionString: process.env.POSTGRES_URL,
});

export const db = drizzle(connection, { schema });
