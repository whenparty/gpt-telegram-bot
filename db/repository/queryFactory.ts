import {
  NodePgDatabase,
  NodePgQueryResultHKT,
} from "drizzle-orm/node-postgres";
import * as schema from "../schema";
import { ExtractTablesWithRelations, InferSelectModel, eq } from "drizzle-orm";
import { PgTransaction } from "drizzle-orm/pg-core";
import { AI_MODEL } from "./aiModels";

export type User = InferSelectModel<(typeof schema)["users"]>;
export type Token = InferSelectModel<(typeof schema)["tokens"]>;

export class QueryFactory {
  constructor(private db: NodePgDatabase<typeof schema>) {}

  getUser(db: NodePgDatabase<typeof schema>, id: number) {
    return db.query.users.findFirst({
      where: eq(schema.users.id, id),
    });
  }

  insertUser(
    tx: PgTransaction<
      NodePgQueryResultHKT,
      typeof schema,
      ExtractTablesWithRelations<typeof schema>
    >,
    user: Omit<User, "id">
  ) {
    return tx
      .insert(schema.users)
      .values({
        externalIdentifier: user.externalIdentifier,
        name: user.name,
        aiModel: AI_MODEL.CLAUDE_3_HAIKU,
      })
      .returning();
  }

  insertTokens(
    tx: PgTransaction<
      NodePgQueryResultHKT,
      typeof schema,
      ExtractTablesWithRelations<typeof schema>
    >,
    userId: number,
    tokens: Omit<Token, "id" | "userId">[]
  ) {
    return tx.insert(schema.tokens).values(
      tokens.map((token) => ({
        ...token,
        userId,
      }))
    );
  }

  updateTokenAmount(
    tx: PgTransaction<
      NodePgQueryResultHKT,
      typeof schema,
      ExtractTablesWithRelations<typeof schema>
    >,
    id: number,
    tokensAmount: number
  ) {
    return tx
      .update(schema.tokens)
      .set({
        tokens: tokensAmount,
      })
      .where(eq(schema.tokens.id, id));
  }
}
