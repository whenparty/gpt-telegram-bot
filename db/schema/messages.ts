import {
  integer,
  pgTable,
  serial,
  char,
  varchar,
  boolean,
  timestamp,
} from "drizzle-orm/pg-core";
import { users } from "./users";

export const messages = pgTable("messages", {
  id: serial("id").primaryKey(),
  role: varchar("role", { enum: ["user", "assistant"] }).notNull(),
  message: varchar("messaage", { length: 4096 }).notNull(),
  sentAt: timestamp("sent_at", { mode: "date" }).notNull().defaultNow(),
  deleted: boolean("deleted").notNull().default(false),
  userId: integer("user_tg_id")
    .notNull()
    .references(() => users.id, {
      onDelete: "cascade",
    }),
});
