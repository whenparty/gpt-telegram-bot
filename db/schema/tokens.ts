import { index, integer, pgTable, serial, unique } from "drizzle-orm/pg-core";
import { users } from "./users";
import { relations } from "drizzle-orm";
import { aiModelEnum } from "./enums";

export const tokens = pgTable(
  "tokens",
  {
    id: serial("id").primaryKey(),
    amount: integer("amount").notNull(),
    aiModel: aiModelEnum("ai_model").notNull(),
    userId: integer("user_id")
      .notNull()
      .references(() => users.id, {
        onDelete: "cascade",
        onUpdate: "restrict",
      }),
  },
  (table) => {
    return {
      userIdx: index("token_user_id_idx").on(table.userId),
      modelUserUk: unique("token_ai_model_user_id_uk").on(
        table.aiModel,
        table.userId
      ),
    };
  }
);

export const tokenRelations = relations(tokens, ({ one }) => ({
  user: one(users, {
    fields: [tokens.userId],
    references: [users.id],
  }),
}));
