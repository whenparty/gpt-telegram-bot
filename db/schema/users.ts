import { relations } from "drizzle-orm";
import { index, pgTable, serial, varchar } from "drizzle-orm/pg-core";
import { tokens } from "./tokens";
import { messages } from "./messages";
import { aiModelEnum } from "./enums";

export const users = pgTable(
  "users",
  {
    id: serial("id").primaryKey(),
    externalIdentifier: varchar("external_identifier", { length: 300 })
      .notNull()
      .unique(),
    name: varchar("name", { length: 300 }).notNull(),
    aiModel: aiModelEnum("ai_model").notNull(),
  },

  (table) => {
    return {
      externalIdentifierIdx: index("external_identifier_idx").on(
        table.externalIdentifier
      ),
    };
  }
);

export const usersRelations = relations(users, ({ many }) => ({
  messages: many(messages),
  tokens: many(tokens),
}));
