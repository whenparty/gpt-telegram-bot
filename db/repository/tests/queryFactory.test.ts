import { it, expect, describe } from "bun:test";
import { QueryFactory } from "../queryFactory";
import { AI_MODEL } from "../aiModels";
import { Client } from "pg";
import { drizzle } from "drizzle-orm/node-postgres";
import * as schema from "db/schema";

describe("Query Factory", () => {
  const db = drizzle(new Client(), { schema });
  const queries = new QueryFactory();

  it("getUser", () => {
    const id = "user_id";
    const { sql, params } = queries.getUserWithTokens(db, id).toSQL();

    const expectedParams = [id, 1];
    expect(params).toStrictEqual(expectedParams);
    expect(sql).toMatchSnapshot(hintFromParams(expectedParams));
  });

  it("insertUser", () => {
    const user = {
      aiModel: AI_MODEL.CLAUDE_3_HAIKU,
      name: "fake",
      externalIdentifier: "fake_id",
    };

    const { sql, params } = queries.insertUser(db as any, user).toSQL();

    const expectedParams = [user.externalIdentifier, user.name, user.aiModel];
    expect(params).toStrictEqual(expectedParams);
    expect(sql).toMatchSnapshot(hintFromParams(expectedParams));
  });

  it("insertTokens", () => {
    const userId = 1;
    const tokens = [
      {
        aiModel: AI_MODEL.CLAUDE_3_HAIKU,
        tokens: 100,
      },
      {
        aiModel: AI_MODEL.CLAUDE_3_OPUS,
        tokens: 200,
      },
    ];

    const { sql, params } = queries
      .insertTokens(db as any, userId, tokens)
      .toSQL();

    const expectedParams = [
      tokens[0].tokens,
      tokens[0].aiModel,
      userId,
      tokens[1].tokens,
      tokens[1].aiModel,
      userId,
    ];
    expect(params).toStrictEqual(expectedParams);
    expect(sql).toMatchSnapshot(hintFromParams(expectedParams));
  });

  it("updateTokenAmount", () => {
    const tokenId = 1;
    const amount = 100;

    const { sql, params } = queries
      .updateTokenAmount(db as any, tokenId, amount)
      .toSQL();

    const expectedParams = [amount, tokenId];
    expect(params).toStrictEqual(expectedParams);
    expect(sql).toMatchSnapshot(hintFromParams(expectedParams));
  });
});

function hintFromParams(params: unknown[]) {
  return params.map((p, i) => `$${i + 1}:${p}`).join(" ");
}
