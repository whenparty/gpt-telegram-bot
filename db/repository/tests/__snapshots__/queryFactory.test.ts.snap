// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`Query Factory getUser: $1:123 $2:1 1`] = `"select "id", "external_identifier", "name", "ai_model" from "users" "userSchema" where "userSchema"."id" = $1 limit $2"`;

exports[`Query Factory insertUser: $1:fake_id $2:fake $3:claude-3-haiku 1`] = `"insert into "users" ("id", "external_identifier", "name", "ai_model") values (default, $1, $2, $3) returning "id", "external_identifier", "name", "ai_model""`;

exports[`Query Factory insertTokens: $1:100 $2:claude-3-haiku $3:1 $4:200 $5:claude-3-opus $6:1 1`] = `"insert into "tokens" ("id", "tokens", "ai_model", "user_id") values (default, $1, $2, $3), (default, $4, $5, $6)"`;

exports[`Query Factory updateTokenAmount: $1:100 $2:1 1`] = `"update "tokens" set "tokens" = $1 where "tokens"."id" = $2"`;
